### PR TITLE
fix: support versions with build metadata in `yarn.lock` files

### LINKED ISSUE #574: Error determining version of yarn dependency with "+git" in version
Saw this today running `osv-scanner` indirectly through Scorecard.
```
Failed to determine version of domino while parsing a yarn.lock - please report this!
```

The offending line is [here](https://github.com/angular/angular/blame/main/packages/zone.js/yarn.lock#L1187), which was added recently [here](
https://github.com/angular/angular/commit/068b38a05d03d50e3051add0aca5ec1cffa2cf4d#diff-b7841b72800c4e42a43ec46745ce2c13253eeb9914dc2fb5cabcea4dfabdb99dR1174)
```
"domino@https://github.com/angular/domino.git#f2435fe1f9f7c91ade0bd472c4723e5eacd7d19a":
  version "2.1.6+git"
  uid f2435fe1f9f7c91ade0bd472c4723e5eacd7d19a
  resolved "https://github.com/angular/domino.git#f2435fe1f9f7c91ade0bd472c4723e5eacd7d19a"
```

I see the `+` symbol is used for build info in [npm semver](https://github.com/npm/node-semver#range-grammar), but I have no idea if this is a valid `yarn.lock` file (similar to #142 )  with any version of yarn. 

Not a `yarn` user, tried a few things and got different (valid) lockfiles. Not sure what Angular did to get theirs. Running `yarn add domino@2.1.6` yields:
```
domino@2.1.6:
  version "2.1.6"
  resolved "https://registry.yarnpkg.com/domino/-/domino-2.1.6.tgz#fe4ace4310526e5e7b9d12c7de01b7f485a57ffe"
  integrity sha512-3VdM/SXBZX2omc9JF9nOPCtDaYQ67BGp5CoLpIQlO2KCAPETs8TcDHacF26jXadGbvUteZzRTeos2fhID5+ucQ==
```

And `yarn add https://github.com/angular/domino.git#f2435fe1f9f7c91ade0bd472c4723e5eacd7d19a` proudces:
```
"@angular/domino@https://github.com/angular/domino.git#f2435fe1f9f7c91ade0bd472c4723e5eacd7d19a":
  version "2.1.6"
  resolved "https://github.com/angular/domino.git#f2435fe1f9f7c91ade0bd472c4723e5eacd7d19a"
```

### PR DESCRIPTION
**Pull request by G-Rath**
Resolves #574

### CODE REVIEW

**Review by another-rex — DISMISSED** [review]
LGTM

File: pkg/lockfile/parse-yarn-lock-v1_test.go [hunk 1]
```diff
@@ -226,6 +226,31 @@ func TestParseYarnLock_v1_VersionsWithBuildString(t *testing.T) {
...
+	packages, err := lockfile.ParseYarnLock("fixtures/yarn/with-build-string.v1.lock")
+
+	if err != nil {
+		t.Errorf("Got unexpected error: %v", err)
+	}
+
+	expectPackages(t, packages, []lockfile.PackageDetails{
+		{
+			Name:      "domino",
+			Version:   "2.1.6+git",
```
Comment (another-rex) [inline pkg/lockfile/parse-yarn-lock-v1_test.go:241]: Should the +git part be trimmed/removed? Or am I misinterpreting this comment: https://github.com/google/osv-scanner/issues/574#issuecomment-1751250600

Comment (G-Rath) [reply pkg/lockfile/parse-yarn-lock-v1_test.go:241]: No it should be retained as its part of the version - its `semantic` that should account for it not being relevant when comparing two versions (which it does).

Comment (another-rex) [reply pkg/lockfile/parse-yarn-lock-v1_test.go:241]: That makes sense, thanks!

**Review by another-rex — APPROVED** [review]

**Pull request merged**